element-hq/element-x-android
# Implement MSC2530

## Type of change


## Content

Based on #2522 (thanks @surakin!) .

- Adds a caption to image and video content when `filename != null && body != filename`.
- Displays the caption below the image or video content in the timeline.
- Changes how images are displayed, following what we suggested some time ago:
  - Behind the image and up until it covers the min size of a bubble a blur hash will be displayed.
  - Over it the image will be scaled until it fits with its aspect ratio, and centered in the available space.
  - We add a white background behind the image to make sure images with alpha channel will display their contents fine.

You can review only my commits, starting from https://github.com/element-hq/element-x-android/commit/6cfe0c85a4132f61969b86e65135db0a81f8ba5d.

## Motivation and context

Closes #2521.

## Screenshots / GIFs

|Image|Video|
|-|-|
|![](https://media.githubusercontent.com/media/element-hq/element-x-android/e2a25dc068f14b1080fc6432f5dc6b9d7a7eb0e5/tests/uitests/src/test/snapshots/images/ui_S_t%5Bf.messages.impl.timeline.components.event_TimelineImageWithCaptionRow_null_TimelineImageWithCaptionRow-Day-37_37_null%2CNEXUS_5%2C1.0%2Cen%5D.png)|![](https://media.githubusercontent.com/media/element-hq/element-x-android/e2a25dc068f14b1080fc6432f5dc6b9d7a7eb0e5/tests/uitests/src/test/snapshots/images/ui_S_t%5Bf.messages.impl.timeline.components.event_TimelineVideoWithCaptionRow_null_TimelineVideoWithCaptionRow-Day-48_48_null%2CNEXUS_5%2C1.0%2Cen%5D.png)|
|![](https://media.githubusercontent.com/media/element-hq/element-x-android/e2a25dc068f14b1080fc6432f5dc6b9d7a7eb0e5/tests/uitests/src/test/snapshots/images/ui_S_t%5Bf.messages.impl.timeline.components.event_TimelineImageWithCaptionRow_null_TimelineImageWithCaptionRow-Night-37_38_null%2CNEXUS_5%2C1.0%2Cen%5D.png)|![](https://media.githubusercontent.com/media/element-hq/element-x-android/e2a25dc068f14b1080fc6432f5dc6b9d7a7eb0e5/tests/uitests/src/test/snapshots/images/ui_S_t%5Bf.messages.impl.timeline.components.event_TimelineVideoWithCaptionRow_null_TimelineVideoWithCaptionRow-Night-48_49_null%2CNEXUS_5%2C1.0%2Cen%5D.png)|

## Tests

You'll need a room where images or videos with captions have been sent.

- Open the room.
- Make sure the images and videos both using and not using captions are displayed fine.

## Tested devices

- OS version(s): 14

## Checklist
